src/frames.ts

```diff
@@ -957,160 +986,140 @@ class RerunnableTask {
-  private _lifecycleCallback: () => void = () => {};
-  private _newDocumentNavigationCompleteCallback: () => void = () => {};
-  private _frameDetachedCallback: (err: Error) => void = () => {};
-  private _navigationAbortedCallback: (err: Error) => void = () => {};
-  private _maximumTimer?: NodeJS.Timer;
-  private _hasSameDocumentNavigation = false;
-  private _targetUrl: string | undefined;
-  private _expectedDocumentId: string | undefined;
-  private _urlMatch: types.URLMatch | undefined;
-
-  constructor(frame: Frame, options: WaitForNavigationOptions | undefined, supportUrlMatch: boolean) {
-    options = options || {};
-    let { waitUntil = 'load' as LifecycleEvent } = options;
-    const { timeout = frame._page._timeoutSettings.navigationTimeout() } = options;
-    if (!Array.isArray(waitUntil))
-      waitUntil = [waitUntil];
-    for (const event of waitUntil) {
-      if (!kLifecycleEvents.has(event))
-        throw new Error(`Unsupported waitUntil option ${String(event)}`);
-    }
-    if (supportUrlMatch)
-      this._urlMatch = options.url;
-    this._expectedLifecycle = waitUntil.slice();
-    this._frame = frame;
-    this.sameDocumentNavigationPromise = new Promise(f => this._sameDocumentNavigationCompleteCallback = f);
-    this.lifecyclePromise = new Promise(f => this._lifecycleCallback = f);
-    this.newDocumentNavigationPromise = new Promise(f => this._newDocumentNavigationCompleteCallback = f);
-    this.timeoutOrTerminationPromise = Promise.race([
-      this._createTimeoutPromise(timeout),
-      new Promise<Error>(f => this._frameDetachedCallback = f),
-      new Promise<Error>(f => this._navigationAbortedCallback = f),
-      this._frame._page._disconnectedPromise.then(() => new Error('Navigation failed because browser has disconnected!')),
-    ]);
-    frame._page._frameManager._lifecycleWatchers.add(this);
-    this._checkLifecycleComplete();
-  }
-
-  _urlMatches(urlString: string): boolean {
-    return !this._urlMatch || platform.urlMatches(urlString, this._urlMatch);
+type Disposable<T> = {value: T, dispose: () => void};
+class Disposer {
+  private _disposes: (() => void)[] = [];
+  add<T>({value, dispose} : Disposable<T>) {
+    this._disposes.push(dispose);
+    return value;
   }
-
-  setExpectedDocumentId(documentId: string, url: string) {
-    assert(!this._urlMatch, 'Should not have url match when expecting a particular navigation');
-    this._expectedDocumentId = documentId;
-    this._targetUrl = url;
-    if (this._navigationRequest && this._navigationRequest._documentId !== documentId)
-      this._navigationRequest = null;
+  dispose() {
+    for (const dispose of this._disposes)
+      dispose();
+    this._disposes = [];
   }
+}
 
-  _onFrameDetached(frame: Frame) {
-    if (this._frame === frame) {
-      this._frameDetachedCallback.call(null, new Error('Navigating frame was detached'));
+function createFrameDestroyedPromise(frame: Frame) : Disposable<Promise<Error>> {
```

Comment:
It would be easier for me to understand methods like `frame._waitForDestroyed` or `frame._waitForNewDocument` instead of these helpers.

Reply:
Done